lib/charms/vault_k8s/v0/vault_autounseal.py

```diff
@@ -0,0 +1,518 @@
+        return _get_credentials_from_secret(secret)
+
+
+def _is_provider_data_valid(data: RelationDataContent) -> bool:
+    """Use the pydantic schema to validate the data."""
+    try:
+        ProviderSchema(app=VaultAutounsealProviderSchema(**data))
+        return True
+    except ValidationError as e:
+        logger.warning("Invalid data: %s", e)
+        return False
+
+
+class VaultAutounsealRequires(ops.Object):
+    """Manages the vault-autounseal relation from the requirer side."""
+
+    on: VaultAutounsealRequireEvents = VaultAutounsealRequireEvents()  # type: ignore
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str):
+        super().__init__(charm, relation_name)
+        self.relation_name = relation_name
+
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+        self.framework.observe(charm.on[relation_name].relation_broken, self._on_relation_broken)
+
+    def _on_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        data = event.relation.data[event.app]
+        if _is_provider_data_valid(data):
+            details = self.get_details()
+            if not details:
+                logger.warning("Missing details, but somehow we passed validation")
+                return
+            self.on.vault_autounseal_details_ready.emit(
+                details.address,
+                details.mount_path,
+                details.key_name,
+                details.role_id,
+                details.secret_id,
+                details.ca_certificate,
+            )
+
+    def _on_relation_broken(self, event: ops.RelationBrokenEvent) -> None:
+        self.on.vault_autounseal_provider_relation_broken.emit()
+
+    def get_details(self) -> AutounsealDetails | None:
+        """Return the vault address, role id, secret id and ca certificate from the relation databag.
+
+        Returns:
+            An AutounsealDetails object if the data is valid, None otherwise.
+        """
+        relation = self.framework.model.get_relation(self.relation_name)
+        if not relation:
+            return None
+        if not relation.active:
+            return None
+        if relation.app is None:
+            logger.warning("No remote application yet")
+            return None
+        data = relation.data[relation.app]
+        address = data.get("address")
+        mount_path = data.get("mount_path")
+        key_name = data.get("key_name")
+        ca_certificate = data.get("ca_certificate")
+        credentials = self._get_credentials(relation)
+        if not (address and mount_path and key_name and ca_certificate and credentials):
+            return None
+        return AutounsealDetails(
+            address,
+            mount_path,
+            key_name,
+            credentials.role_id,
+            credentials.secret_id,
+            ca_certificate,
+        )
+
+    def _get_credentials(self, relation: ops.Relation) -> ApproleDetails | None:
```

Comment:
This is the same code as in L389

Reply:
Those are in different classes and the common code seems to be abstracted into `_get_credentials_from_secret`

Reply:
L389
```
    def _get_credentials(self, relation: ops.Relation) -> ApproleDetails | None:
        """Retrieve the credentials from the Juju secret.

        Args:
            relation: The relation to get the credentials for.

        Returns:
            An ApproleDetails object if the credentials are found, None otherwise.
        """
        if not relation.active:
            logger.warning("Relation is not active")
            return None
        if relation.app is None:
            logger.warning("No remote application yet")
            return None
        credentials_secret_id = relation.data[self.charm.app].get("credentials_secret_id")
        if credentials_secret_id is None:
            return None
        secret = self.model.get_secret(id=credentials_secret_id)
        return _get_credentials_from_secret(secret)
```

L483
```
    def _get_credentials(self, relation: ops.Relation) -> ApproleDetails | None:
        """Return the token from the Juju secret.

        Returns:
            A tuple containing the role id and secret id
        """
        if not relation.active:
            logger.warning("Relation is not active")
            return None
        if relation.app is None:
            logger.warning("No remote application yet")
            return None
        credentials_secret_id = relation.data[relation.app].get("credentials_secret_id")
        if not credentials_secret_id:
            return None
        secret = self.model.get_secret(id=credentials_secret_id)
        return _get_credentials_from_secret(secret)
```